tiledb/sm/query/ast/query_ast.cc

```diff
@@ -0,0 +1,143 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Implements Query AST classes.
+ */
+
+#include "query_ast.h"
+
+namespace tiledb {
+namespace sm {
+
+tdb_unique_ptr<ASTNode> ast_combine(
+    const tdb_unique_ptr<ASTNode>& lhs,
+    const tdb_unique_ptr<ASTNode>& rhs,
+    QueryConditionCombinationOp combination_op) {
+  // AST Construction.
+  std::vector<tdb_unique_ptr<ASTNode>> ast_nodes;
+  if (lhs->get_tag() == ASTNodeTag::VAL) {
+    if (rhs->get_tag() == ASTNodeTag::VAL) {
+      ast_nodes.push_back(lhs->clone());
+      ast_nodes.push_back(rhs->clone());
+    } else {
+      // lhs is a simple tree, rhs is a compound tree.
+      auto rhs_tree_expr = dynamic_cast<ASTNodeExpr*>(rhs.get());
+      ast_nodes.push_back(lhs->clone());
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    }
+  } else if (rhs->get_tag() == ASTNodeTag::VAL) {
+    // lhs is a compound tree, rhs is a simple tree.
+    auto lhs_tree_expr = dynamic_cast<ASTNodeExpr*>(lhs.get());
+    for (const auto& elem : lhs_tree_expr->nodes_) {
+      ast_nodes.push_back(elem->clone());
+    }
+    ast_nodes.push_back(rhs->clone());
+  } else if (
+      lhs->get_tag() == ASTNodeTag::EXPR &&
+      rhs->get_tag() == ASTNodeTag::EXPR) {
+    // Both trees are expression trees.
+    auto lhs_tree_expr = dynamic_cast<ASTNodeExpr*>(lhs.get());
+    auto rhs_tree_expr = dynamic_cast<ASTNodeExpr*>(rhs.get());
+
+    if (combination_op == lhs_tree_expr->combination_op_ &&
+        lhs_tree_expr->combination_op_ == rhs_tree_expr->combination_op_) {
+      // The combination ops of the lhs compound tree, rhs compound tree, and
+      // the combination op argument passed in are identical.
+      for (const auto& elem : lhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else if (combination_op == lhs_tree_expr->combination_op_) {
+      // The combination op argument passed in is the same as the combination op
+      // of the lhs compound tree.
+      for (const auto& elem : lhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+      ast_nodes.push_back(rhs->clone());
+    } else if (combination_op == rhs_tree_expr->combination_op_) {
+      // The combination op argument passed in is the same as the combination op
+      // of the rhs compound tree.
+      ast_nodes.push_back(lhs->clone());
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else {
+      ast_nodes.push_back(lhs->clone());
+      ast_nodes.push_back(rhs->clone());
+    }
```

Comment:
I think you can do this whole function with something like
```c++
  if (combination_op == lhs_tree_expr->combination_op_()) {
      for (const auto& elem : lhs_tree_expr->nodes_) {
         ast_nodes.push_back(elem->clone());
       }
  } else {
    ast_nodes.push_back(lhs->clone());
  } 
   if (combination_op == rhs_tree_expr->combination_op_()) {
      for (const auto& elem : rhs_tree_expr->nodes_) {
         ast_nodes.push_back(elem->clone());
       }
  }  else {
    ast_nodes.push_back(rhs->clone());
  } 
```
where `combination_op_()` is a virtual function that returns the op (including non_op for a value)

Reply:
Wait true, thanks!